rust/lance-encoding/src/compression.rs

```diff
@@ -468,6 +480,35 @@ impl CompressionStrategy for DefaultCompressionStrategy {
         match data {
             DataBlock::FixedWidth(_) => Ok(Box::new(ValueEncoder::default())),
             DataBlock::FixedSizeList(_) => Ok(Box::new(ValueEncoder::default())),
+            DataBlock::Struct(struct_block) => {
+                if field.children.len() != struct_block.children.len() {
+                    return Err(Error::invalid_input(
+                        "Struct field metadata does not match data block children",
+                        location!(),
+                    ));
+                }
+                let has_variable_child = struct_block
+                    .children
+                    .iter()
+                    .any(|child| !matches!(child, DataBlock::FixedWidth(_)));
+                if has_variable_child {
+                    if self.version < LanceFileVersion::V2_2 {
+                        return Err(Error::NotSupported {
+                            source: "Variable packed struct encoding requires Lance file version 2.2 or later".into(),
+                            location: location!(),
+                        });
+                    }
+                    Ok(Box::new(PackedStructVariablePerValueEncoder::new(
+                        self.clone(),
+                        field.children.clone(),
+                    )))
+                } else {
+                    Err(Error::invalid_input(
+                        "Packed struct per-value compression should not be used for fixed-width-only structs",
+                        location!(),
+                    ))
```

Comment:
This can be in a follow-up but we probably will want some kind of per-value solution for fixed-width data types.  The current mini-block solution is somewhat limited and not great if there are many fields.

Reply:
Yep, I have the same feeling, created https://github.com/lancedb/lance/issues/5021 to follow.